### ISSUE TITLE
Make SecretProviderClass Namespaced scoped

### ISSUE THREAD

**Issue by ritazh** [kubernetes-sigs/secrets-store-csi-driver#60]

**Comment by munnerz** [comment]
I'm new to the project and trying to get my head around how cert-manager could also build a provider for this project (to replace/unify efforts between this and https://github.com/jetstack/cert-manager-csi, which appears to have similar goals).

Given that CSIDrivers themselves are cluster-scoped, how do you see scoping this to a single namespace working? I can see the benefits, and cert-manager itself has a concept of an 'Issuer' vs a 'ClusterIssuer' (representing a 'source' of certificates/a CA), and we could make use of namespaced SecretProviderClasses ourselves to expose our Issuer resources. But I'm not too sure on a practical/isolation/boundaries point of view how this can be exposed if a SecretProviderClass roughly maps to a CSIDriver 🤔

**Comment by ritazh** [comment]
@munnerz I think that's a great idea! For more details on how to add a new provider, please refer to: https://github.com/kubernetes-sigs/secrets-store-csi-driver#providers

> Given that CSIDrivers themselves are cluster-scoped, how do you see scoping this to a single namespace working?

The idea here is within the csi driver, instead of relying on its own service account to get all SecretProviderClasses resources, we would then use the namespace and service account of the requesting pod to get to the namespaced SecretProviderClasses resources it has access to. WDYT?

**Comment by helayoty** [comment]
May I start working on this issue?

**Comment by aramase** [comment]
@helayoty Sure, go for it! 

/assign @helayoty

**Comment by aramase** [comment]
Add new field in the `volumeAttributes` for providing the `secretProviderClassNamespace`.

```yaml
        volumeAttributes:
          secretProviderClass: "azure-kvname-podid"
          secretProviderClassNamespace: "ns1"
```

If the `secretProviderClassNamespace` is provided, then filter on that ns, if not use the pod ns to filter the `SecretProviderClass`.

**Comment by aramase** [comment]
As part of the new controller for syncing secrets, a part of this change has already been implemented - https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/pkg/secrets-store/utils.go#L163-L166. The `SecretProviderClass` is being filtered by namespace. So the remaining change that's required as part of this issue is 

1. To read `secretProviderClassNamespace` from `volumeAttributes`
2. If present, then pass this namespace to the function [here](https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/pkg/secrets-store/utils.go#L155) instead of the default pod namespace that's being passed.

**Comment by aramase** [comment]
Closing this issue as the `SecretProviderClass` in matched in the same namespace as pod with https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/224/files#diff-a31c6f6257e17c698bbc44d4dbb22a7dR124. After discussion on the community call, I've opened an issue #242 to gather user feedback on if this is a feature they would like to see. Once we have +1s, we can rethink about `secretProviderClassNamespace`. 

/close
